Realm/RLMObjectSchema.mm

```diff
@@ -123,6 +123,13 @@ + (instancetype)schemaForObjectClass:(Class)objectClass {
         }
     }
 
+    for (RLMProperty *prop in schema.properties) {
+        RLMPropertyType type = prop.type;
+        if (prop.optional && type != RLMPropertyTypeObject && type != RLMPropertyTypeArray && type != RLMPropertyTypeString) {
+            @throw RLMException([NSString stringWithFormat:@"Only 'string', 'array', and 'object' properties can be made optional, and property '%@' is of type '%@'.", prop.name, RLMTypeToString(type)]);
```

Comment:
Optional arrays aren't supported by core at the moment and I don't think there's really plans to change that.